src/main/java/org/assertj/core/internal/Files.java

```diff
@@ -88,10 +89,15 @@ public static Files instance() {
   public void assertSameContentAs(AssertionInfo info, File actual, File expected) {
     verifyIsFile(expected);
     assertIsFile(info, actual);
+    BinaryDiffResult binaryDiffResult = null;
     try {
+      binaryDiffResult = binaryDiff.diff(actual, java.nio.file.Files.readAllBytes(expected.toPath()));
```

Comment:
I agree with you, It's not possible to have different binary content and equal text content.

The global algorithm is: 
```
If 
 binary equivalent: return ok
else
 compare line by line
 if line by line comparison is not possible, return binary comparison
 return the differences, human readable
```

So in this case the `diff` is only used to return the most readable thing possible:
 1. the line by line diff
 2. the binary diff
 3. the `Unable to compare contents of files` error

For the ```java
public S hasSameContentAs(File expected, Charset charset)
```
I can add it but I'd rather do it in another PR, what do you think ?